R/geom-defaults.r

```diff
@@ -15,6 +15,7 @@ update_geom_defaults <- function(geom, new) {
   g <- check_subclass(geom, "Geom", env = parent.frame())
   old <- g$default_aes
   g$default_aes <- defaults(rename_aes(new), old)
+  class(g$default_aes) <- "uneval"
```

Comment:
Perhaps the following is more foolproof if, for whatever reason, somebody creates a geom where the defaults don't follow conventions.
```r
g$default_aes[] <- defaults(rename_aes(new), old)
```
This runs the result through `ggplot2:::new_aes()` whenever `g$default_aes` is a proper mapping with the uneval class. You could then leave out the manual class assignment.

This also applies to the stat variant.

Reply:
I like the idea of preserving the order is nice. However, I think there are some niche but legitimate use-cases for defining aesthetics outside the already available aesthetics. Using the `[names(old)]` would ensure that such aesthetics are dropped, so I would be more comfortable if you could revert that particular change.

Reply:
The problem is that the other strategy doesn't actually work because it mixes aesthetics:

``` r
library("ggplot2") # using g$default_aes[] <- defaults(rename_aes(new), old)

GeomPoint$default_aes
#> Aesthetic mapping: 
#> * `shape`  -> 19
#> * `colour` -> "black"
#> * `size`   -> 1.5
#> * `fill`   -> NA
#> * `alpha`  -> NA
#> * `stroke` -> 0.5

update_geom_defaults("point", list(color = "red"))

GeomPoint$default_aes
#> Aesthetic mapping: 
#> * `shape`  -> "red"
#> * `colour` -> 19
#> * `size`   -> 1.5
#> * `fill`   -> NA
#> * `alpha`  -> NA
#> * `stroke` -> 0.5

ggplot(mtcars, aes(disp, mpg)) + geom_point()
#> Error in `geom_point()`:
#> ! Problem while converting geom to grob.
#> ℹ Error occurred in the 1st layer.
#> Caused by error in `translate_shape_string()` at ggplot2/R/geom-point.r:122:6:
#> ! Shape aesthetic contains invalid value: "red"
```

<sup>Created on 2023-01-17 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

For defining aesthetics outside the already available aesthetics, I think the `g$default_aes[] <- ` will bug then regardless of how we do the naming, right?, because there will be more new default aesthetics than the old list holds. Does that mean back to the `class() <- "uneval"`? (Possibly after reordering, and appending new aesthetics?)

Reply:
Good catch! Yeah the uneval class method could work, but that also wouldn't preserve the order I think. We might also make the renaming of `new` a separate step, then we could use those to preserve order. Something like the following:

``` r
# Path might differ on your machine, but this is to use all the
# internals
devtools::load_all("~/packages/ggplot2")
#> ℹ Loading ggplot2

update_geom_defaults2 <- function(geom, new) {
  g <- check_subclass(geom, "Geom", env = parent.frame())
  old <- g$default_aes
  new <- rename_aes(new)
  new <- defaults(new, old)[union(names(old), names(new))]
  g$default_aes[names(new)] <- new
  invisible()
}

GeomPoint$default_aes
#> Aesthetic mapping: 
#> * `shape`  -> 19
#> * `colour` -> "black"
#> * `size`   -> 1.5
#> * `fill`   -> NA
#> * `alpha`  -> NA
#> * `stroke` -> 0.5

update_geom_defaults2("point", list(color = "red"))

GeomPoint$default_aes
#> Aesthetic mapping: 
#> * `shape`  -> 19
#> * `colour` -> "red"
#> * `size`   -> 1.5
#> * `fill`   -> NA
#> * `alpha`  -> NA
#> * `stroke` -> 0.5

update_geom_defaults2("point", list(foo = "bar"))

GeomPoint$default_aes
#> Aesthetic mapping: 
#> * `shape`  -> 19
#> * `colour` -> "red"
#> * `size`   -> 1.5
#> * `fill`   -> NA
#> * `alpha`  -> NA
#> * `stroke` -> 0.5
#> * `foo`    -> "bar"
```

<sup>Created on 2023-01-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
I'm fine with that as well, but I think the difference in readability is pretty small.